### ISSUE TITLE
Tasks running endless times with DaskExecutor when working on variables that take a lot of memory

### ISSUE THREAD

**Issue by neumann-nico** [PrefectHQ/prefect#2772]
## Description
I'm working on reading big files and processing them. When executing prefect in asynchronous mode everything works well. When I'm trying to run the task in parallel with DaskExecutor I get a strange behavior. If a variable takes a lot of memory, all the tasks start to run in an endless loop until it crashes.

Printing hello1, hello2, hello3 should only happen once. But you can see that it gets printed many times until it crashes.

```bash
[2020-06-12 13:23:08] INFO - prefect.FlowRunner | Beginning Flow run for 'parallel-execution'
[2020-06-12 13:23:08] INFO - prefect.FlowRunner | Starting flow run.
[2020-06-12 13:23:09] INFO - prefect.TaskRunner | Task 'print_message': Starting task run...
hello1
[2020-06-12 13:23:09] INFO - prefect.TaskRunner | Task 'print_message': Starting task run...
hello3
[2020-06-12 13:23:09] INFO - prefect.TaskRunner | Task 'print_message': Starting task run...
hello2
[2020-06-12 13:23:09] INFO - prefect.TaskRunner | Task 'create_dict': Starting task run...
[2020-06-12 13:23:09] INFO - prefect.TaskRunner | Task 'print_message': finished task run for task with final state: 'Success'
[2020-06-12 13:23:09] INFO - prefect.TaskRunner | Task 'print_message': finished task run for task with final state: 'Success'
[2020-06-12 13:23:09] INFO - prefect.TaskRunner | Task 'print_message': finished task run for task with final state: 'Success'
[2020-06-12 13:23:15] INFO - prefect.TaskRunner | Task 'create_dict': Starting task run...
[2020-06-12 13:23:21] INFO - prefect.TaskRunner | Task 'print_message': Starting task run...
hello1
[2020-06-12 13:23:21] INFO - prefect.TaskRunner | Task 'print_message': finished task run for task with final state: 'Success'
[2020-06-12 13:23:21] INFO - prefect.TaskRunner | Task 'create_dict': Starting task run...
[2020-06-12 13:23:27] INFO - prefect.TaskRunner | Task 'create_dict': Starting task run...
[2020-06-12 13:23:33] INFO - prefect.TaskRunner | Task 'print_message': Starting task run...
hello3
[2020-06-12 13:23:33] INFO - prefect.TaskRunner | Task 'print_message': finished task run for task with final state: 'Success'
[2020-06-12 13:23:33] INFO - prefect.TaskRunner | Task 'print_message': Starting task run...
hello1
[2020-06-12 13:23:33] INFO - prefect.TaskRunner | Task 'print_message': finished task run for task with final state: 'Success'
[2020-06-12 13:23:35] ERROR - prefect.FlowRunner | Unexpected error: KilledWorker('create_dict-138a4290-3d35-4213-bba2-f106bbfec807', <Worker 'tcp://127.0.0.1:38291', name: 3, memory: 0, processing: 1>)
Traceback (most recent call last):
  File "/home/nico/git/datachain/venv/lib/python3.7/site-packages/prefect/engine/runner.py", line 48, in inner
    new_state = method(self, state, *args, **kwargs)
  File "/home/nico/git/datachain/venv/lib/python3.7/site-packages/prefect/engine/flow_runner.py", line 463, in get_flow_run_state
    for t in final_tasks
  File "/home/nico/git/datachain/venv/lib/python3.7/site-packages/prefect/engine/executors/dask.py", line 321, in wait
    return self.client.gather(futures)
  File "/home/nico/git/datachain/venv/lib/python3.7/site-packages/distributed/client.py", line 1974, in gather
    asynchronous=asynchronous,
  File "/home/nico/git/datachain/venv/lib/python3.7/site-packages/distributed/client.py", line 824, in sync
    self.loop, func, *args, callback_timeout=callback_timeout, **kwargs
  File "/home/nico/git/datachain/venv/lib/python3.7/site-packages/distributed/utils.py", line 339, in sync
    raise exc.with_traceback(tb)
  File "/home/nico/git/datachain/venv/lib/python3.7/site-packages/distributed/utils.py", line 323, in f
    result[0] = yield future
  File "/home/nico/git/datachain/venv/lib/python3.7/site-packages/tornado/gen.py", line 735, in run
    value = future.result()
  File "/home/nico/git/datachain/venv/lib/python3.7/site-packages/distributed/client.py", line 1833, in _gather
    raise exception.with_traceback(traceback)
distributed.scheduler.KilledWorker: ('create_dict-138a4290-3d35-4213-bba2-f106bbfec807', <Worker 'tcp://127.0.0.1:38291', name: 3, memory: 0, processing: 1>)
[2020-06-12 13:23:35] ERROR - prefect.parallel-execution | Unexpected error occured in FlowRunner: KilledWorker('create_dict-138a4290-3d35-4213-bba2-f106bbfec807', <Worker 'tcp://127.0.0.1:38291', name: 3, memory: 0, processing: 1>)
```

## Expected Behavior

The task should run once and then finish.

```bash
[2020-06-12 13:20:44] INFO - prefect.FlowRunner | Beginning Flow run for 'parallel-execution'
[2020-06-12 13:20:44] INFO - prefect.FlowRunner | Starting flow run.
[2020-06-12 13:20:44] INFO - prefect.TaskRunner | Task 'print_message': Starting task run...
hello1
[2020-06-12 13:20:44] INFO - prefect.TaskRunner | Task 'print_message': finished task run for task with final state: 'Success'
[2020-06-12 13:20:44] INFO - prefect.TaskRunner | Task 'print_message': Starting task run...
hello2
[2020-06-12 13:20:44] INFO - prefect.TaskRunner | Task 'print_message': finished task run for task with final state: 'Success'
[2020-06-12 13:20:44] INFO - prefect.TaskRunner | Task 'print_message': Starting task run...
hello3
[2020-06-12 13:20:44] INFO - prefect.TaskRunner | Task 'print_message': finished task run for task with final state: 'Success'
[2020-06-12 13:20:44] INFO - prefect.TaskRunner | Task 'create_dict': Starting task run...
[2020-06-12 13:20:53] INFO - prefect.TaskRunner | Task 'create_dict': finished task run for task with final state: 'Success'
[2020-06-12 13:20:53] INFO - prefect.FlowRunner | Flow run SUCCESS: all reference tasks succeeded
```

## Reproduction
If I run the same code without DaskExecutor `flow.run()` the process finishes without problems.
If you run the code with DaskExecutor and reduce the dict by deleting some zeros in the loop the task gets executed successfully too. For me changing 1000 to 100 did the job.

```python
#!/usr/bin/python

from prefect import Flow, Parameter, task
from prefect.engine.executors import DaskExecutor


@task
def print_message(message):
    print(message)


@task
def create_dict():
    my_dict = dict()

    for i in range(100):
        my_dict[i] = dict()
        for j in range(500):
            my_dict[i][j] = dict()
            for k in range(1000):
                my_dict[i][j][k] = 0

    return my_dict


def main():
    with Flow('parallel-execution') as flow:
        print_message('hello1')
        print_message('hello2')
        print_message('hello3')

        my_dict = create_dict()

    state = flow.run(executor=DaskExecutor())


if __name__ == '__main__':
    main()
```

## Environment
```bash
{
  "config_overrides": {},
  "env_vars": [],
  "system_information": {
    "platform": "Linux-4.15.0-65-generic-x86_64-with-Ubuntu-16.04-xenial",
    "prefect_version": "0.11.5",
    "python_version": "3.7.7"
  }
}
```

**Comment by cicdw** [comment]
Hi @neumann-nico - this is happening because your "workers" (which in this case are being created under-the-hood) are running out of their allocated memory.   Occasionally when a worker dies, Dask needs to recompute tasks that were lost (hence the endless prints). You can recreate this without Prefect:

```python
from distributed import Client

def create_dict():
    my_dict = dict()

    for i in range(100):
        my_dict[i] = dict()
        for j in range(500):
            my_dict[i][j] = dict()
            for k in range(1000):
                my_dict[i][j][k] = 0

    return my_dict

with Client() as client:
    client.gather(client.submit(create_dict))
```

To fix your issue, I recommend standing up the workers yourself via the dask CLI (commands all run in different terminal windows):
```
>> dask-scheduler

## start two workers
>> dask-worker tcp://10.0.0.41:8786
>> dask-worker tcp://10.0.0.41:8786
```

You can read more about Dask worker memory management here: https://distributed.dask.org/en/latest/worker.html#memory-management

**Comment by casey-green** [comment]
In case anyone else runs into this issue, I was able to work around it by changing from `DaskExecutor` to `LocalDaskExecutor(scheduler="threads")`. I'm **assuming** this is because the large task results no longer needed to be de/serialized across worker processes, which likely takes additional memory. This is not a sure-fire workaround, as you may still run out of memory if your task results are large enough.

Moving from processes to threads is an OK workaround for us because our tasks are more I/O bound, and not very CPU intensive.
